test/SgidClient.spec.ts

```diff
@@ -0,0 +1,332 @@
+      const expected = new URL(MOCK_AUTH_ENDPOINT)
+      expect(actual.host).toBe(expected.host)
+      expect(actual.pathname).toBe(expected.pathname)
+      expect(actual.searchParams.get('client_id')).toBe(MOCK_CLIENT_ID)
+      expect(actual.searchParams.get('scope')).toBe(DEFAULT_SCOPE)
+      expect(actual.searchParams.get('response_type')).toBe(
+        DEFAULT_RESPONSE_TYPE,
+      )
+      // Important check is here
+      expect(actual.searchParams.get('redirect_uri')).toBe(mockRedirectUri)
+      expect(actual.searchParams.get('nonce')).toBe(nonce)
+      expect(actual.searchParams.get('state')).toBe(mockState)
+      // Client ID, scope, response_type, redirect_uri, nonce, state
+      expect(actualNumSearchParams).toBe(6)
+    })
+
+    it('should throw when no redirectUri is provided', () => {
+      const mockState = 'mockState'
+
+      const client = new SgidClient({
+        clientId: MOCK_CLIENT_ID,
+        clientSecret: MOCK_CLIENT_SECRET,
+        privateKey: MOCK_CLIENT_PRIVATE_KEY,
+        hostname: MOCK_HOSTNAME,
+        apiVersion: MOCK_API_VERSION,
+      })
+
+      expect(() => client.authorizationUrl(mockState)).toThrow(
+        'No redirect URI registered with this client',
+      )
+    })
+  })
+
+  describe('callback', () => {
+    it('should call token endpoint and return sub and accessToken', async () => {
+      const { sub, accessToken } = await client.callback(MOCK_AUTH_CODE)
+
+      expect(sub).toBe(MOCK_SUB)
+      expect(accessToken).toBe(MOCK_ACCESS_TOKEN)
+    })
+
+    it('should throw when no access token is returned', async () => {
+      server.use(tokenHandlerNoToken)
+
+      await expect(client.callback(MOCK_AUTH_CODE)).rejects.toThrow(
+        'Missing sub claim or access token',
+      )
+    })
+
+    it('should throw when sub is empty', async () => {
+      server.use(tokenHandlerNoSub)
+
+      await expect(client.callback(MOCK_AUTH_CODE)).rejects.toThrow(
+        'Missing sub claim or access token',
+      )
+    })
+  })
+
+  describe('userinfo', () => {
+    it('should call userinfo endpoint and return sub and data', async () => {
+      const { sub, data } = await client.userinfo(MOCK_ACCESS_TOKEN)
+
+      expect(sub).toBe(MOCK_SUB)
+      expect(data).toEqual(MOCK_USERINFO_PLAINTEXT)
+    })
+
+    it('should return empty data object when no key is returned', async () => {
+      server.use(userInfoHandlerNoKey)
+
+      const { sub, data } = await client.userinfo(MOCK_ACCESS_TOKEN)
+
+      expect(sub).toBe(MOCK_SUB)
+      expect(data).toEqual({})
+    })
+
+    it('should return empty data object when no data is returned', async () => {
+      server.use(userInfoHandlerNoData)
+
+      const { sub, data } = await client.userinfo(MOCK_ACCESS_TOKEN)
+
+      expect(sub).toBe(MOCK_SUB)
+      expect(data).toEqual({})
+    })
+
+    it('should throw when private key is invalid', async () => {
+      const client = new SgidClient({
```

Comment:
same as above comment. rename client.

```suggestion
      const invalidClient = new SgidClient({
```

Reply:
[a8f7374](https://github.com/opengovsg/sgid-client/pull/24/commits/a8f737474b342dca211714bb454fa727f299ebf4)